File: server/src/routes/serial.ts
import type { FastifyInstance } from 'fastify';
import { SerialPort } from 'serialport';
import type { SerialPortsResponse } from '@muxus/shared';

interface ListedSerialPort {
  path: string;
  manufacturer?: string;
  serialNumber?: string;
  pnpId?: string;
  locationId?: string;
  productId?: string;
  vendorId?: string;
}

export function serialPortsResponse(ports: readonly ListedSerialPort[]): SerialPortsResponse {
  return {
    ports: ports
      .map((port) => ({
        path: port.path,
        manufacturer: port.manufacturer,
        serialNumber: port.serialNumber,
        pnpId: port.pnpId,
        locationId: port.locationId,
        productId: port.productId,
        vendorId: port.vendorId,
      }))
      .sort((a, b) => a.path.localeCompare(b.path, undefined, { numeric: true })),
  };
}

/** Enumerate OS serial devices for the saved-host editor. */
export function registerSerialRoutes(app: FastifyInstance): void {
  app.get('/api/serial/ports', async (): Promise<SerialPortsResponse> => {
    return serialPortsResponse(await SerialPort.list());
  });
}

File: app/internal/api/serial.go
package api

// SerialPortInfo mirrors SerialPortInfo: one serial device reported by the
// host OS. Optional fields use omitempty because the Node server serializes
// missing values as absent keys (undefined is dropped by JSON.stringify).
type SerialPortInfo struct {
	// Path is the OS-native path (COM3, /dev/ttyUSB0, /dev/tty.usbserial-…).
	Path         string `json:"path"`
	Manufacturer string `json:"manufacturer,omitempty"`
	SerialNumber string `json:"serialNumber,omitempty"`
	PnpID        string `json:"pnpId,omitempty"`
	LocationID   string `json:"locationId,omitempty"`
	ProductID    string `json:"productId,omitempty"`
	VendorID     string `json:"vendorId,omitempty"`
}

// SerialPortsResponse mirrors SerialPortsResponse.
type SerialPortsResponse struct {
	Ports []SerialPortInfo `json:"ports"`
}
